### PR TITLE
fix(limits): Count concurrent PR with platform API instead of branches

### LINKED ISSUE #907: Renovate says prsRemaining: -4 when 0 PRs are open
### Which Renovate are you using?

Renovate Open Source CLI

### Which platform are you using?

GitHub.com

### Have you checked the logs? Don't forget to include them if relevant

Yes. I see the following, even though there are exactly 0 renovate PRs open:

```json
DEBUG: PR hourly limit remaining: 3
DEBUG: Enforcing prConcurrentLimit (3)
DEBUG: 7 PRs are currently open
DEBUG: PR concurrent limit remaining: -4
DEBUG: Calculated maximum PRs remaining this run
{
  "prsRemaining": -4
}
```

### What would you like to do?

Have renovate open PRs for these branches. Renovate has 7 branches created, but no PRs. Here is my `renovate.json`:

```json
{
  "extends": ["config:base", ":automergeMinor"],
  "labels": ["renovate"],
  "dependencyDashboard": true,
  "packageRules": [
    {
      "updateTypes": ["major"],
      "dependencyDashboardApproval": true
    },
    {
      "extends": "packages:linters",
      "groupName": "linters"
    }
  ],
  "postUpdateOptions": ["yarnDedupeHighest"],
  "prBodyColumns": ["Package", "Update", "Type", "Change", "References"],
  "prConcurrentLimit": 3,
  "prCreation": "not-pending",
  "prHourlyLimit": 3,
  "rebaseWhen": "behind-base-branch",
  "semanticCommits": true,
  "stabilityDays": 3
}
```

### PR DESCRIPTION
**Pull request by zharinov**
Closes https://github.com/renovatebot/config-help/issues/907

### CODE REVIEW

**Review by viceice — DISMISSED** [review]

File: lib/workers/repository/process/limits.ts [hunk 1]
```diff
@@ -40,22 +40,32 @@ export async function getPrHourlyRemaining(
...
-        currentlyOpen += 1;
-      }
+    logger.debug(`Calculating prConcurrentLimit (${config.prConcurrentLimit})`);
+    try {
+      const branchList = branches.map(({ branchName }) => branchName);
+      const prList = await Promise.all(
+        branchList.map((branchName) => platform.getBranchPr(branchName))
+      );
+      const openPrs = prList
+        .filter(Boolean)
```
Comment (viceice) [inline lib/workers/repository/process/limits.ts:55]: 🤔  Do we really need this?

Comment (zharinov) [reply lib/workers/repository/process/limits.ts:55]: Not sure about "really" part of your question, but `getBranchPr` can possibly return `null` according to its signature. To make it more bulletproof, I would even replace `Promise.all` with manual loop with try-catch checks.

Comment (viceice) [reply lib/workers/repository/process/limits.ts:55]: ok, it shouldn't really matter if we do this sequencially, as the pr list would be fetch only once from server.

**Comment by viceice** [comment]
request review when you're done. 🙃

File: lib/workers/repository/process/limits.ts (hunk 1 above)
Comment (rarkins) [reply lib/workers/repository/process/limits.ts:55]: I think this logic makes sense to filter out any `null` responses for when PRs don't exist for the branch yet

**Comment by zharinov** [comment]
Now it's ready for review

**Comment by viceice** [comment]
@zharinov  next time click the rounded arrows to rerequest review :wink:
image

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
Awaiting real repo tests

**Review by rarkins — APPROVED** [review]

File: lib/workers/repository/process/write.ts [hunk 2]
```diff
@@ -64,6 +76,14 @@ export async function writeUpdates(
     ) {
       deductPrRemainingCount = 1;
     }
+    // istanbul ignore if
```
Comment (viceice) [inline lib/workers/repository/process/write.ts:79]: This should be tested?

Comment (zharinov) [reply lib/workers/repository/process/write.ts:79]: I tested it manually, most probably this and other checks above will go

Comment (viceice) [reply lib/workers/repository/process/write.ts:79]: ok, we should add test later

**Review by viceice — APPROVED** [review]

**Pull request merged**